File: scripts/pdf_extract.js
// JXA PDF Extractor
ObjC.import('PDFKit');
ObjC.import('Foundation');

function run(argv) {
    if (argv.length < 1) {
        console.log("Usage: osascript -l JavaScript pdf_extract.js <input_pdf>");
        return;
    }

    var inputPath = argv[0];
    var url = $.NSURL.fileURLWithPath(inputPath);
    var doc = $.PDFDocument.alloc.initWithURL(url);

    if (doc.isNil()) {
        return;
    }

    var pageCount = doc.pageCount;
    var fullText = "";

    for (var i = 0; i < pageCount; i++) {
        var page = doc.pageAtIndex(i);
        var text = page.string.js;
        if (text) {
            fullText += text + "\n";
        }
    }

    // Return text to stdout
    return fullText;
}

File: scripts/pdf_extract.swift
#!/usr/bin/swift

import Foundation
import PDFKit

// PDF Text Extractor
// Usage: swift pdf_extract.swift <input_pdf_path> <output_txt_path>

guard CommandLine.arguments.count == 3 else {
    print("Usage: swift pdf_extract.swift <input_pdf> <output_txt>")
    exit(1)
}

let inputPath = CommandLine.arguments[1]
let outputPath = CommandLine.arguments[2]

let url = URL(fileURLWithPath: inputPath)

guard let document = PDFDocument(url: url) else {
    print("Error: Could not load PDF at \(inputPath)")
    exit(1)
}

// Check if document is encrypted
if document.isEncrypted {
    print("Warning: Document is encrypted. Extraction might fail.")
}

var fullText = ""
let pageCount = document.pageCount

print("Extracting text from \(pageCount) pages...")

for i in 0..<pageCount {
    if let page = document.page(at: i) {
        if let pageText = page.string {
            fullText += pageText + "\n"
        }
    }
}

do {
    try fullText.write(to: URL(fileURLWithPath: outputPath), atomically: true, encoding: .utf8)
    print("✅ Extracted text to \(outputPath)")
    print("   Length: \(fullText.count) characters")
} catch {
    print("Error writing output: \(error)")
    exit(1)
}
